### PR TITLE
Add Optimism+Arbitrum Implementation

### PR DESCRIPTION
**Pull request by hexonaut**
Adding support for both optimism and arbitrum to dss-bridge.

Note this is using new feature multichain forking of Foundry so be sure to do a `foundryup`.

### CODE REVIEW

File: packages/dss-bridge/src/domains/optimism/OptimismDomainHost.sol [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        uint32 gasLimit;
+        assembly {
+            gasLimit := mload(loc)
+        }
+        if (gasLimit == 0) gasLimit = glLift;
+
+        l1messenger.sendMessage(
+            guest,
+            abi.encodeWithSelector(DomainGuest.lift.selector, id, line, minted),
+            glLift
```
Comment (olivdb) [inline packages/dss-bridge/src/domains/optimism/OptimismDomainHost.sol:82]: I think you meant `gasLimit` (not `glLift`) here

**Comment by hexonaut** [comment]
To be clear this is still WIP, but want comments before proceeding.

File: packages/dss-bridge/src/domains/optimism/OptimismDomainHost.sol [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        l1messenger = L1MessengerLike(_l1messenger);
+        guest = _guest;
+    }
+
+    // Example of providing custom parameters to the bridge
+    function lift(uint256 wad, uint32 gasLimit) external {
+        // Stash extra args in a magic memory location
+        bytes32 loc = MAGIC_MEM_LOC;
+        assembly {
+            mstore(loc, gasLimit)
```
Comment (olivdb) [inline packages/dss-bridge/src/domains/optimism/OptimismDomainHost.sol:61]: This feels a bit too contrived IMO. Why not simply inheriting the common code via an internal function like so:

```
abstract contract DomainHost {
    function _lift(uint256 wad) internal auth vatLive {
        uint256 rad = wad * RAY;
        ...

        line = rad;

        emit Lift(wad);
    }
}

contract OptimismDomainHost {
    function lift(uint256 wad, uint32 gasLimit) external {
        _lift(wad);

        if (gasLimit == 0) gasLimit = glLift;
        l1messenger.sendMessage(
            guest,
            abi.encodeWithSelector(DomainGuest.lift.selector, id, line, minted),
            gasLimit
        );
    }
}
```

**Comment by hexonaut** [comment]
Updated to better design with no callbacks.

**Comment by hexonaut** [comment]
Actually nvm there is more changes I want to do.

**Comment by hexonaut** [comment]
Ok now this is ready for review after merging the l2 messaging into `dss-test`.

**Comment by hexonaut** [comment]
Added Arbitrum support as well.

File: packages/dss-bridge/src/domains/arbitrum/ArbitrumDomainGuest.sol [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity ^0.8.14;
+
+import {DomainGuest,TeleportGUID} from "../../DomainGuest.sol";
+import {DomainHost} from "../../DomainHost.sol";
```
Comment (gbalabasquer) [inline packages/dss-bridge/src/domains/arbitrum/ArbitrumDomainGuest.sol:23]: I don't think we should import this one.

File: packages/dss-bridge/src/domains/arbitrum/ArbitrumDomainHost.sol [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity ^0.8.14;
+
+import {DomainHost,TeleportGUID} from "../../DomainHost.sol";
+import {DomainGuest} from "../../DomainGuest.sol";
```
Comment (gbalabasquer) [inline packages/dss-bridge/src/domains/arbitrum/ArbitrumDomainHost.sol:23]: Same here. I see they are used for the selectors, but shouldn't be necessary to expose them in the main file (I think).

**Comment by gbalabasquer** [comment]
So to understand a bit better the new structure, `ArbitrumDomainHost/Guest` and `OptimismDomainHost/Guest` would be the only contracts managing bridge interactions (for Teleport, regular bridging and MCD actions).
That would mean the packages of `arbitrum-dai-bridge` and `optimism-dai-bridge` would end up being deprecated, right?
If that's the case, shouldn't maybe just move `dss-bridge` as a different (and just solidity) repo?

**Comment by hexonaut** [comment]
Yes I've been thinking the same thing of moving to a separate repo. Originally I thought there would be dependencies in the domain-specific repos, but it's becoming clear this is not the case.

Do you want to move to a new repo right away or wait until this PR is reviewed?

**Comment by gbalabasquer** [comment]
> Yes I've been thinking the same thing of moving to a separate repo. Originally I thought there would be dependencies in the domain-specific repos, but it's becoming clear this is not the case.
> 
> Do you want to move to a new repo right away or wait until this PR is reviewed?

I think we can get these PRs merged first then move to new repos.

**Comment by hexonaut** [comment]
Going to merge this into the parent PR.

**Pull request merged**